src/main/telemetry/smartport.c

```diff
@@ -160,30 +169,97 @@ static uint8_t smartPortHasRequest = 0;
 
+typedef struct smartPortFrame_s {
+    uint8_t  sensorId;
+    uint8_t  frameId;
+    uint16_t valueId;
+    uint32_t data;
+} smartPortFrame_t;
+
+#define SMARTPORT_FRAME_SIZE  sizeof(smartPortFrame_t)
+#define SMARTPORT_TX_BUF_SIZE 256
+
+#define SMARTPORT_PAYLOAD_SIZE   (SMARTPORT_FRAME_SIZE - 2*sizeof(uint8_t))
+#define SMARTPORT_PAYLOAD_OFFSET 2
+
+static uint8_t smartPortRxBuffer[SMARTPORT_FRAME_SIZE];
+static uint8_t smartPortRxBytes = 0;
+static bool smartPortFrameReceived = false;
+
+#define SMARTPORT_MSP_VERSION    1
+#define SMARTPORT_MSP_VER_SHIFT  5
+#define SMARTPORT_MSP_VER_MASK   0xE0
+#define SMARTPORT_MSP_VERSION_S  (SMARTPORT_MSP_VERSION << SMARTPORT_MSP_VER_SHIFT)
+
+#define SMARTPORT_MSP_START_FLAG (1 << 4)
+#define SMARTPORT_MSP_SEQ_MASK   0x0F
+
+#define SMARTPORT_MSP_RX_BUF_SIZE 64
+
+static uint8_t smartPortMspTxBuffer[SMARTPORT_TX_BUF_SIZE];
+static mspPacket_t smartPortMspReply;
+static bool smartPortMspReplyPending = false;
+
 static void smartPortDataReceive(uint16_t c)
 {
+    static bool skipUntilStart = true;
+    static bool byteStuffing = false;
+
     uint32_t now = millis();
 
-    // look for a valid request sequence
-    static uint8_t lastChar;
-    if (lastChar == FSSP_START_STOP) {
-        smartPortState = SPSTATE_WORKING;
-        if (c == FSSP_SENSOR_ID1 && (serialRxBytesWaiting(smartPortSerialPort) == 0)) {
+    if (c == FSSP_START_STOP) {
+        smartPortRxBytes = 0;
+        smartPortHasRequest = 0;
+        skipUntilStart = false;
+        return;
+    }
+    else if (skipUntilStart) {
+        return;
+    }
+
+    if (smartPortRxBytes == 0) {
+        if ((c == FSSP_SENSOR_ID1) && (serialRxBytesWaiting(smartPortSerialPort) == 0)) {
+
+            // our slot is starting...
             smartPortLastRequestTime = now;
             smartPortHasRequest = 1;
-            // we only responde to these IDs
-            // the X4R-SB does send other IDs, we ignore them, but take note of the time
+        }
+        else if (c == FSSP_SENSOR_ID2) {
+            smartPortRxBuffer[smartPortRxBytes++] = c;
+        }
+        else {
+            skipUntilStart = true;
+        }
+    }
+    else {
+
+        //TODO: add CRC checking
+
+        if (c == FSSP_DLE) {
+            byteStuffing = true;
+            return;
+        }
+
+        if (byteStuffing) {
+            c ^= FSSP_DLE_XOR;
+            byteStuffing = false;
+        }
+
+        smartPortRxBuffer[smartPortRxBytes++] = c;
+
+        if(smartPortRxBytes == SMARTPORT_FRAME_SIZE) {
+            smartPortFrameReceived = true;
+            skipUntilStart = true;
         }
     }
-    lastChar = c;
 }
 
 static void smartPortSendByte(uint8_t c, uint16_t *crcp)
 {
     // smart port escape sequence
-    if (c == 0x7D || c == 0x7E) {
-        serialWrite(smartPortSerialPort, 0x7D);
-        c ^= 0x20;
+    if (c == FSSP_DLE || c == FSSP_START_STOP) {
+        serialWrite(smartPortSerialPort, FSSP_DLE);
+        c ^= FSSP_DLE_XOR;
```

Comment:
Now I got your point! Sorry for the confusion. I change it to not modify "c", so that the CRC is computed on original, and not on escaped value.